### ISSUE TITLE
False positive with pytest.skip

### ISSUE THREAD

**Issue by NeilGirdhar** [PyCQA/pylint#6815]
### Bug description

```python
@pytest.fixture(scope='session', params=create_infos())
def distribution_info(request: Any) -> Optional[DistributionInfo[Any, Any, Any]]:
    distribution_name: Optional[str] = request.config.getoption('--distribution')
    info_name = type(request.param).__name__[:-4]
    if distribution_name is None or info_name == distribution_name:
        return request.param
    pytest.skip(f"Deselected {info_name}")
```
Produces
```
Useless suppression of 'inconsistent-return-statements' (useless-suppression)
```
even though `pytest.skip` is marked `NoReturn`.

### Command used

```shell
.
```

### Pylint output

```shell
.
```

### Expected behavior

.

### Pylint version

```shell
pylint 2.14.0
astroid 2.11.5
Python 3.10.0 (default, Nov  6 2021, 07:41:13) [Clang 12.0.5 (clang-1205.0.22.9)]
```

**Comment by NeilGirdhar** [comment]
@DanielNoord Sorrry I meant 
```
tests/conftest.py:45:0: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
```
My mistake.

**Comment by DanielNoord** [comment]
Since this is related to `pylint` I would recommend opening this against either [pylint-pytest](https://pypi.org/project/pylint-pytest/) or [pytest-pylint](https://pypi.org/project/pytest-pylint/).

Those are community maintained extensions for pylint which offer support for pytest. Pylint itself doesn't offer this.

**Comment by NeilGirdhar** [comment]
@DanielNoord [I'm not using ](https://github.com/NeilGirdhar/efax/blob/main/pyproject.toml) either of those extensions.  Isn't pylint just using its own analysis to evaluate this error?

**Comment by DanielNoord** [comment]
> @DanielNoord [I'm not using ](https://github.com/NeilGirdhar/efax/blob/main/pyproject.toml) either of those extensions. Isn't pylint just using its own analysis to evaluate this error?

Yes, but for some external packages we need to rely on some hints to make sure we understand the packages correctly. However, we can't do this for all packages, as we are also limited in our maintainers' time. That's why some external packages are supported through community plugins, like [`pylint-django`](https://pypi.org/project/pylint-django/). The same goes for `pytest`.

**Comment by NeilGirdhar** [comment]
> Yes, but for some external packages we need to rely on some hints to make sure we understand the packages correctly

I understand, but isn't the type annotation enough?

**Comment by NeilGirdhar** [comment]
Here's an example that doesn't use pytest:
```python
from typing import Any, Callable, Protocol, Type, TypeVar, cast

_F = TypeVar("_F", bound=Callable[..., object])
_ET = TypeVar("_ET", bound=Type[BaseException])


class _WithException(Protocol[_F, _ET]):
    Exception: _ET
    __call__: _F


def _with_exception(exception_type: _ET) -> Callable[[_F], _WithException[_F, _ET]]:
    def decorate(func: _F) -> _WithException[_F, _ET]:
        func_with_exception = cast(_WithException[_F, _ET], func)
        func_with_exception.Exception = exception_type
        return func_with_exception

    return decorate


@_with_exception(BaseException)
def skip(reason: str = "") -> "NoReturn":
    print(reason)


def distribution_info(request: Any):
    if request is not None:  # dummy condition
        return None
    skip("Deselected")
```
@DanielNoord Would you mind reopening this issue?

**Comment by Pierre-Sassoulas** [comment]
This is not a false positive, the issue here is that the return are inconsistent in the function. I.e. `` all return statements in a function should return an expression, or none of them should``, if you return ``request.param`` you should also return everywhere else. This should work.
```python
from typing import Any

import pytest


def distribution_info(request: Any):
    if request is not None:  # dummy condition
        return request.param
    return pytest.skip("Deselected {info_name}")

```    
See https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/inconsistent-return-statements.html?highlight=inconsistent-return

**Comment by NeilGirdhar** [comment]
@Pierre-Sassoulas  I believe you're mistaken.  `pytest.skip` is marked `NoReturn`, therefore, it is exempt from the rule that you "should also return elsewhere".  Your fix doesn't make any sense either since returning from a `NoReturn` function doesn't make sense.

One fix that does work is changing the annotation in Pytest from `"NoReturn"` to `NoReturn` as I did in my linked pull request.  Another fix would be for Pylint to understand text annotations.

**Comment by Pierre-Sassoulas** [comment]
Well adding a return is easier to understand for both human and linter imo, it's just a clue that it's going to "return" without having to check what skip does so you just have to read the function to understand it easily. If you think this check and the fix it currently require does not make sense you can disable it.

We have an issue for taking typing into account https://github.com/PyCQA/pylint/issues/4813, it's not going to be in pylint before a long time. pylint is not mypy, mypy has 20k open issues, and pylint has it's own issues to take care off.

**Comment by NeilGirdhar** [comment]
> Well adding a return is easier to understand for both human and linter imo

I respectfully disagree.  When calling a no return function, you should not return.  What you're suggesting amounts to writing code like this:
```python
def f(x: int) -> bool:
    if x < 0:
        raise ValueError
        return False
```
Adding `return False` doesn't make the code "easier to understand" for humans, and it shouldn't affect linters.

>  without having to check what skip does so you just have to read the function to understand it easily.

I understand what you're saying, but I don't think it helps.

> We have an issue for taking typing into account https://github.com/PyCQA/pylint/issues/4813, it's not going to be in pylint before a long time. pylint is not mypy, mypy has 2k+ open issues, and pylint has it's own issues to take care off.

Fair enough!  Let's leave this closed then as low priority, and hopefully Pytest accepts my pull request, which will resolve this issue for me.
